File: classfields/api-tests/salesman-api-client/src/main/scala/ru/auto/tests/provider/RequestSpecSupplier.scala
package ru.auto.tests.provider

import io.restassured.builder.RequestSpecBuilder
import io.restassured.config.LogConfig
import io.restassured.config.RestAssuredConfig
import ru.auto.tests.commons.restassured.AllureFailureListener
import ru.auto.tests.commons.restassured.AllureLoggerFilter
import java.net.URI
import io.restassured.config.FailureConfig.failureConfig
import io.restassured.config.LogConfig.logConfig
import io.restassured.config.ObjectMapperConfig.objectMapperConfig
import io.restassured.config.RestAssuredConfig.config
import io.restassured.http.ContentType.JSON
import ru.auto.tests.GsonObjectMapper.gson

object RequestSpecSupplier {

  private[provider] def baseReqSpecBuilder(baseUri: URI, basePath: String) =
    new RequestSpecBuilder()
      .setContentType(JSON)
      .setBaseUri(baseUri)
      .setBasePath(basePath)
      .addFilter(new AllureLoggerFilter)

  private[provider] def getLogConfig(loggerEnabled: Boolean): LogConfig = {
    if (loggerEnabled)
      return logConfig.enableLoggingOfRequestAndResponseIfValidationFails
    logConfig
  }

  private[provider] def baseRestAssuredConfig =
    config
      .objectMapperConfig(objectMapperConfig.defaultObjectMapper(gson))
      .failureConfig(failureConfig.failureListeners(new AllureFailureListener))
}

class RequestSpecSupplier private () {}

File: classfields/api-tests/public-api-client/src/main/java/ru/auto/tests/publicapi/provider/RequestSpecSupplier.java
package ru.auto.tests.publicapi.provider;

import io.restassured.builder.RequestSpecBuilder;
import io.restassured.config.LogConfig;
import io.restassured.config.RestAssuredConfig;
import org.apache.commons.codec.binary.Hex;
import ru.auto.tests.commons.restassured.AllureFailureListener;
import ru.auto.tests.publicapi.ra.AllureLoggerFilter;
import ru.auto.tests.publicapi.ra.InternalServerErrorListener;

import java.net.URI;
import java.util.Random;

import static io.restassured.config.FailureConfig.failureConfig;
import static io.restassured.config.LogConfig.logConfig;
import static io.restassured.config.ObjectMapperConfig.objectMapperConfig;
import static io.restassured.config.RestAssuredConfig.config;
import static io.restassured.http.ContentType.JSON;
import static ru.auto.tests.publicapi.GsonObjectMapper.gson;
import static ru.auto.tests.publicapi.ra.RequestSpecBuilders.UBER_TRACE_ID;

class RequestSpecSupplier {

    private RequestSpecSupplier() {

    }

    static RequestSpecBuilder baseReqSpecBuilder(URI baseUri, String basePath) {
        return new RequestSpecBuilder().setContentType(JSON)
                .setBaseUri(baseUri)
                .setBasePath(basePath)
                .addFilter(new AllureLoggerFilter())
                .addHeader(UBER_TRACE_ID, getUberTraceId());
    }

    private static String getUberTraceId() {
        byte[] bytes = new byte[8];
        new Random().nextBytes(bytes);
        String traceId = Hex.encodeHexString(bytes);
        return String.format("%s:%s:0:1", traceId, traceId);
    }

    static LogConfig getLogConfig(boolean loggerEnabled) {
        if (loggerEnabled) {
            return logConfig().enableLoggingOfRequestAndResponseIfValidationFails();
        }
        return logConfig();
    }

    static RestAssuredConfig baseRestAssuredConfig() {
        return config().objectMapperConfig(objectMapperConfig().defaultObjectMapper(gson()))
                .failureConfig(failureConfig().failureListeners(new AllureFailureListener(), new InternalServerErrorListener()));
    }
}
